Offer a Meteor 1.4-compatible vagrant-spk platform stack

Context:
- If you try to package an app that uses Meteor 1.4.x with vagrant-spk's Meteor platform stack, you will run into mysterious build problems.
- This is due to version incompatibilities between the nodejs version bundled with Meteor pre-1.4 and the node version bundled with Meteor 1.4+.
- A while ago, @zarvox and I discussed this, and the best solution we could come up with was to have two different platform stacks -- one for Meteor 1.4+ and one for Meteor pre-1.4.

Sub-problems to consider:
- vagrant-spk should properly package a Meteor apps that has no `.sandstorm/`. This may involve looking at  `.meteor/release` and printing an error message, or auto-switching which `.sandstorm/` it generates.
- vagrant-spk should provide reasonable advice for apps that used to Meteor 1.3 and now are using Meteor 1.4. This may involve looking at  `.meteor/release` and printing an link to docs.

FWIW here's the output with current meteor-spk & Meteor 1.4 code:

```
paulproteus@slittingmill:~/projects/meteor-example-clock$ vagrant-spk dev
Calling 'vagrant' 'ssh' '-c' '/opt/app/.sandstorm/build.sh && cd /opt/app/.sandstorm && spk dev --pkg-def=/opt/app/.sandstorm/sandstorm-pkgdef.capnp:pkgdef' in /home/paulproteus/projects/meteor-example-clock/.sandstorm
npm WARN package.json meteor-dev-bundle@0.0.0 No description                       
npm WARN package.json meteor-dev-bundle@0.0.0 No repository field.
npm WARN package.json meteor-dev-bundle@0.0.0 No README data
 
> fibers@1.0.13 install /home/vagrant/bundle/programs/server/node_modules/fibers
> node build.js || nodejs build.js

make: Entering directory '/home/vagrant/bundle/programs/server/node_modules/fibers/build'
  CXX(target) Release/obj.target/fibers/src/fibers.o
make: g++: Command not found
fibers.target.mk:91: recipe for target 'Release/obj.target/fibers/src/fibers.o' failed
make: *** [Release/obj.target/fibers/src/fibers.o] Error 127
make: Leaving directory '/home/vagrant/bundle/programs/server/node_modules/fibers/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/vagrant/.meteor/packages/meteor-tool/.1.1.3.4sddkj++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System Linux 3.16.0-4-amd64
gyp ERR! command "node" "/home/vagrant/.meteor/packages/meteor-tool/.1.1.3.4sddkj++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild" "--release"
gyp ERR! cwd /home/vagrant/bundle/programs/server/node_modules/fibers
gyp ERR! node -v v0.10.36
gyp ERR! node-gyp -v v1.0.1
gyp ERR! not ok 
node-gyp exited with code: 1
Please make sure you are using a supported platform and node version. If you
would like to compile fibers on this machine please make sure you have setup your
build environment--
Windows + OS X instructions here: https://github.com/nodejs/node-gyp
Ubuntu users please run: `sudo apt-get install g++`
Alpine users please run: `sudo apk add python make g++`
sh: 1: nodejs: not found

npm ERR! fibers@1.0.13 install: `node build.js || nodejs build.js`
npm ERR! Exit status 127
npm ERR! 
npm ERR! Failed at the fibers@1.0.13 install script.
npm ERR! This is most likely a problem with the fibers package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node build.js || nodejs build.js
npm ERR! You can get their info via:
npm ERR!     npm owner ls fibers
npm ERR! There is likely additional logging output above.
npm ERR! System Linux 3.16.0-4-amd64
npm ERR! command "/home/vagrant/.meteor/packages/meteor-tool/.1.1.3.4sddkj++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/node" "/home/vagrant/.meteor/packages/meteor-tool/.1.1.3.4sddkj++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/bin/npm" "install"
npm ERR! cwd /home/vagrant/bundle/programs/server
npm ERR! node -v v0.10.36
npm ERR! npm -v 1.4.28
npm ERR! code ELIFECYCLE
npm ERR! not ok code 0
Connection to 127.0.0.1 closed.
Traceback (most recent call last):
  File "/usr/local/bin/vagrant-spk", line 1028, in <module>
    main()
  File "/usr/local/bin/vagrant-spk", line 1025, in main
    operation(args)
  File "/usr/local/bin/vagrant-spk", line 680, in dev
    "spk dev --pkg-def=/opt/app/.sandstorm/sandstorm-pkgdef.capnp:pkgdef"
  File "/usr/local/bin/vagrant-spk", line 294, in call_vagrant_command
    return subprocess.check_call(command, cwd=sandstorm_dir)
  File "/usr/lib/python2.7/subprocess.py", line 540, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['vagrant', 'ssh', '-c', '/opt/app/.sandstorm/build.sh && cd /opt/app/.sandstorm && spk dev --pkg-def=/opt/app/.sandstorm/sandstorm-pkgdef.capnp:pkgdef']' returned non-zero exit status 1
```

Things I've noticed so far, when trying a Sandstorm app based on Meteor 1.4:

- The Meteor platform stack needs a C++ compiler, which is harmless to add for all versions of Meteor.

    - Note, however, that this is a change to `setup.sh` that existing app authors would need to make.

- Meteor likes to hang forever at "Updating package catalog" and/or create an invalid sqlite database when you ^C it, which results in a pretty sad state of affairs. This _seems_ to be fixed by the following change in `setup.sh`:

```
diff --git a/.sandstorm/setup.sh b/.sandstorm/setup.sh
index 8bdc63e..d6e07e2 100755
--- a/.sandstorm/setup.sh
+++ b/.sandstorm/setup.sh
@@ -42,7 +42,7 @@ cp -a /lib/x86_64-linux-gnu/libtinfo.so.* /opt/meteor-spk/meteor-spk.deps/lib/x8
 # Unfortunately, Meteor does not explicitly make it easy to cache packages, but
 # we know experimentally that the package is mostly directly extractable to a
 # user's $HOME/.meteor directory.
-METEOR_RELEASE=1.1.0.2
+METEOR_RELEASE=1.4.1.2
 METEOR_PLATFORM=os.linux.x86_64
 METEOR_TARBALL_FILENAME="meteor-bootstrap-${METEOR_PLATFORM}.tar.gz"
 METEOR_TARBALL_URL="https://d3sqy0vbqsdhku.cloudfront.net/packages-bootstrap/${METEOR_RELEASE}/${METEOR_TARBALL_FILENAME}"
```

    - It's possible that using the newer Meteor release allows us to overcome a Meteor CDN bug that @kentonv mentioned this morning. This does, however, pin us to a particular version of Meteor. I'd prefer to always download the latest release of Meteor, which we can theoretically get by doing `curl https://install.meteor.com/ | grep ^RELEASE=` but that's pretty ugly.

    - This is also a change that existing app authors probably need to do, to get past the weird CDN problem. It's probably safe to pin this to a newer version for all Meteor-based Sandstorm app authors. It's probably unsafe to make this totally dynamic, since that decreases how "reproducible" the vagrant-spk builds are, so the above is probably a bad idea.

- The app works properly if I set `PACKAGE=meteor-spk-0.3.0`. Interestingly, if I set `PACKAGE=meteor-spk-0.1.8` the build "succeeds" but I get this stuff in my grain log:

```
...** SANDSTORM SUPERVISOR: Starting up grain.
** Starting Mongo...
2016-10-06T23:03:26.831+0000 I STORAGE  Engine custom option: log=(prealloc=false,file_max=200KB)
about to fork child process, waiting until server is ready for connections.
forked process: 6
child process started successfully, parent exiting
** Starting Meteor...

assert.js:93
  throw new assert.AssertionError({
        ^
AssertionError: "undefined" === "function"
    at wrapPathFunction (/programs/server/mini-files.js:77:10)
    at Object.<anonymous> (/programs/server/mini-files.js:108:24)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Module.require (module.js:364:17)
    at require (module.js:380:17)
    at Object.<anonymous> (/programs/server/boot.js:9:13)
    at Module._compile (module.js:456:26)
sandstorm/sandstorm-http-bridge.c++:2236: warning: App isn't listening for TCP connections after 30 seconds. Continuing to attempt to connect; address->toString() = 127.0.0.1:8000
```

--- a/stacks/meteor/build.sh
+++ b/stacks/meteor/build.sh
@@ -3,12 +3,19 @@ set -euo pipefail
 
 # Make meteor bundle
 
-METEOR_WAREHOUSE_DIR="${METEOR_WAREHOUSE_DIR:-/home/vagrant/.meteor}"
-METEOR_DEV_BUNDLE=$(dirname $(readlink -f "$METEOR_WAREHOUSE_DIR/meteor"))/dev_bundle
+export NODE_ENV=production
+sudo chown vagrant:vagrant /home/vagrant -R
 
 cd /opt/app
+meteor npm install --production
 meteor build --directory /home/vagrant/
-(cd /home/vagrant/bundle/programs/server && "$METEOR_DEV_BUNDLE/bin/npm" install)
+# Use npm and node from the Meteor dev bundle to install the bundle's dependencies.
+TOOL_VERSION=$(meteor show --ejson $(<.meteor/release) | grep '^ *"tool":' |
+    sed -re 's/^.*"(meteor-tool@[^"]*)".*$/\1/g')
+TOOLDIR=$(echo $TOOL_VERSION | tr @ /)
+PATH=$HOME/.meteor/packages/$TOOLDIR/mt-os.linux.x86_64/dev_bundle/bin:$PATH
+cd /home/vagrant/bundle/programs/server
+npm install --production
 
 # Copy our launcher script into the bundle so the grain can start up.
 mkdir -p /home/vagrant/bundle/opt/app/.sandstorm/

--- a/stacks/meteor/setup.sh
+++ b/stacks/meteor/setup.sh
@@ -6,10 +6,13 @@
 set -euo pipefail
 
 CURL_OPTS="--silent --show-error"
+apt-get update
+apt-get install -y build-essential git
 
 cd /opt/
 
-PACKAGE=meteor-spk-0.1.8
+NODE_ENV=production
+PACKAGE=meteor-spk-0.4.1
 PACKAGE_FILENAME="$PACKAGE.tar.xz"
 CACHE_TARGET="/host-dot-sandstorm/caches/${PACKAGE_FILENAME}"
 
@@ -36,7 +39,7 @@ cp -a /lib/x86_64-linux-gnu/libtinfo.so.* /opt/meteor-spk/meteor-spk.deps/lib/x8
 # Unfortunately, Meteor does not explicitly make it easy to cache packages, but
 # we know experimentally that the package is mostly directly extractable to a
 # user's $HOME/.meteor directory.
-METEOR_RELEASE=1.1.0.2
+METEOR_RELEASE=1.6.1.1
 METEOR_PLATFORM=os.linux.x86_64
 METEOR_TARBALL_FILENAME="meteor-bootstrap-${METEOR_PLATFORM}.tar.gz"
 METEOR_TARBALL_URL="https://d3sqy0vbqsdhku.cloudfront.net/packages-bootstrap/${METEOR_RELEASE}/${METEOR_TARBALL_FILENAME}"
@@ -55,3 +58,4 @@ cd /home/vagrant/
 su -c "tar xf '${METEOR_CACHE_TARGET}'" vagrant
 # Link into global PATH
 if [ ! -e /usr/bin/meteor ] ; then ln -s /home/vagrant/.meteor/meteor /usr/bin/meteor ; fi
+chown vagrant:vagrant /home/vagrant -R

